txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+                        return Err(Error::NotInsertedIOWrongAmount.into());
+                    }
+                }
+                Output::Variable { to, amount, color } => {
+                    if is_output_filled {
+                        if to != i_owner {
+                            return Err(Error::NotInsertedIOWrongOwner.into());
+                        }
+                        if amount < i_amount {
+                            return Err(Error::NotInsertedIOWrongAmount.into());
+                        }
+                        if color != i_color {
+                            return Err(Error::NotInsertedIOWrongAssetIt.into());
+                        }
+                    }
+                    // else do nothing, everything is variable and can be only check on execution
+                }
+                Output::ContractCreated { .. } => {
+                    return Err(Error::NotInsertedIOConractOutput.into())
+                }
+            };
+        } else {
+            panic!("Use it only for coin output check");
+        }
+        Ok(())
+    }
+
+    /// insert tx inside dependency
+    /// return list of transactions that are removed from txpool
+    pub async fn insert<'a>(
+        &'a mut self,
+        txs: &'a HashMap<TxId, ArcTx>,
+        db: &dyn TxPoolDB,
+        tx: &'a ArcTx,
+    ) -> anyhow::Result<Vec<ArcTx>> {
+        let mut collided: Vec<TxId> = Vec::new();
+        // iterate over all inputs and check for colision
+        let mut max_depth = 0;
+        let mut db_coins: HashMap<UtxoId, CoinState> = HashMap::new();
+        let mut db_contracts: HashMap<ContractId, ContractState> = HashMap::new();
+        for input in tx.inputs() {
+            // check if all required inputs are here.
+            match input {
+                Input::Coin { utxo_id, .. } => {
+                    // is it dependent output?
+                    if let Some(state) = self.coins.get_mut(utxo_id) {
+                        // check depth
+                        max_depth = core::cmp::max(state.depth + 1, max_depth);
+                        if max_depth > self.max_depth {
+                            return Err(Error::NotInsertedMaxDepth.into());
+                        }
+                        // output is present but is it spend by other tx?
+                        if let Some(ref spend_by) = state.is_spend_by {
+                            // get tx that is spending this output
+                            let txpool_tx = txs
+                                .get(spend_by)
+                                .expect("Tx should be always present in txpool");
+                            // compare if tx has better price
+                            if txpool_tx.gas_price() > tx.gas_price() {
+                                return Err(Error::NotInsertedCollision(*spend_by, *utxo_id).into());
+                            } else {
+                                if state.depth == 0 {
```

Comment:
Shouldn't we check if the tx depth if it's already in the db / included in a block before comparing gas prices?

Reply:
Comaprison does apply only for tx in txpool. This comparison is between transaction that collided on same output `spend_by` contains transaction in txpool that currently spends that output.